docs: add reference docs and guides for custom resources (#334)
Kong/kubernetes-ingress-controller: docs/guides/using-kongingress-resource.md

# Using KongIngress resource

In this guide, we will learn how to use KongIngress resource to control
proxy behavior.

## Installation

Please follow the [deployment](../deployment) documentation to install
Kong Ingress Controller onto your Kubernetes cluster.

## Testing connectivity to Kong

This guide assumes that the `PROXY_IP` environment variable is
set to contain the IP address or URL pointing to Kong.
Please follow one of the
[deployment guides](../deployment) to configure this environment variable.

If everything is setup correctly, making a request to Kong should return
HTTP 404 Not Found.

```bash
$ curl -i $PROXY_IP
HTTP/1.1 404 Not Found
Date: Fri, 21 Jun 2019 17:01:07 GMT
Content-Type: application/json; charset=utf-8
Connection: keep-alive
Content-Length: 48
Server: kong/1.2.1

{"message":"no Route matched with those values"}
```

> This is expected since Kong doesn't know how to proxy the request yet.

Reviewer: ```suggestion
This is expected as Kong does not yet know how to proxy the request.
```

# Using KongIngress resource

In this guide, we will learn how to use KongIngress resource to control
proxy behavior.

## Installation

Please follow the [deployment](../deployment) documentation to install
Kong Ingress Controller onto your Kubernetes cluster.

## Testing connectivity to Kong

> This guide assumes that `PROXY_IP` environment varialbe is

Reviewer: ```suggestion
This guide assumes that the `PROXY_IP` environment variable is
```

set to contain the IP address or URL pointing to Kong.

> If you've not done so, please follow one of the

Reviewer: ```suggestion
Please follow one of the
```

[deployment guides](../deployment) to configure this environment variable.

> If everything is setup correctly, making a request to Kong should return back

Reviewer: ```suggestion
If everything is setup correctly, making a request to Kong should return
```

> a HTTP 404 Not Found.

Reviewer: ```suggestion
HTTP 404 Not Found.
```

```bash
$ echo "
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  name: demo
spec:
  rules:
  - http:
      paths:
      - path: /foo
        backend:
          serviceName: echo
          servicePort: 80
" | kubectl apply -f -
ingress.extensions/demo created
```

Let's test:

```bash
$ curl -i $PROXY_IP/foo
HTTP/1.1 200 OK
Content-Type: text/plain; charset=UTF-8
Transfer-Encoding: chunked
Connection: keep-alive
Date: Wed, 17 Jul 2019 21:38:17 GMT
Server: echoserver
X-Kong-Upstream-Latency: 2
X-Kong-Proxy-Latency: 1
Via: kong/1.2.1

Hostname: echo-d778ffcd8-n9bss

Pod Information:
	node name:	gke-harry-k8s-dev-default-pool-bb23a167-8pgh
	pod name:	echo-d778ffcd8-n9bss
	pod namespace:	default
	pod IP:	10.60.0.4

Server values:
	server_version=nginx: 1.12.2 - lua: 10010

Request Information:
	client_address=10.60.1.10
	method=GET
	real path=/
	query=
	request_version=1.1
	request_scheme=http
	request_uri=http://35.233.170.67:8080/
<-- clipped -- >
```

## Use KongIngress with Ingress resource

Kong will strip the path defined in the Ingress rule before proxying
the request to the service.
This can be seen in the real path value in the above response.

> Let's say, we would like to configure Kong to not do that and also

Reviewer: ```suggestion
We can configure Kong to not strip out this path and to only respond to GET requests
```

> respond to only GET request for this particular Ingress rule.

Reviewer: ```suggestion
for this particular Ingress rule.
```

service/echo patched
```

Let's test this now:

```bash
$ curl $PROXY_IP/foo/hello-world
Hostname: echo-d778ffcd8-vrrtw

Pod Information:
	node name:	gke-harry-k8s-dev-default-pool-bb23a167-8pgh
	pod name:	echo-d778ffcd8-vrrtw
	pod namespace:	default
	pod IP:	10.60.0.9

Server values:
	server_version=nginx: 1.12.2 - lua: 10010

Request Information:
	client_address=10.60.1.10
	method=GET
	real path=/bar/foo/hello-world
	query=
	request_version=1.1
	request_scheme=http
	request_uri=http://35.233.170.67:8080/bar/foo/hello-world

<-- clipped -->
```

Real path received by the upstream service (echo) is now changed to `/bar/foo`.

Also, now all the requests will be sent to the same upstream pod:

```bash
$ curl -s $PROXY_IP/foo/hello-world | grep "pod IP"
	pod IP:	10.60.0.9
$ curl -s $PROXY_IP/foo/hello-world | grep "pod IP"
	pod IP:	10.60.0.9
$ curl -s $PROXY_IP/foo/hello-world | grep "pod IP"
	pod IP:	10.60.0.9
$ curl -s $PROXY_IP/foo/hello-world | grep "pod IP"
	pod IP:	10.60.0.9
$ curl -s $PROXY_IP/foo/hello-world | grep "pod IP"
	pod IP:	10.60.0.9
$ curl -s $PROXY_IP/foo/hello-world | grep "pod IP"
	pod IP:	10.60.0.9
```

> And that's it for this guide.

Reviewer: ```suggestion
```
